### PR TITLE
Firmware

### PR DESCRIPTION
**Pull request by hdiomede**
### Description

Adds Firmware driver resource implementation and integrate firmware bundle with it in the same CHEF resource
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass (`$ rake test`).
  - [ ] New resources and/or actions have are included in the [matchers.rb](https://github.com/HewlettPackard/oneview-chef/blob/master/libraries/matchers.rb) and [matchers_spec](https://github.com/HewlettPackard/oneview-chef/blob/master/spec/unit/resources/matchers_spec.rb).
- [ ] New functionality has been documented in the README if applicable.
  - [ ] New functionality has been thoroughly documented in [examples](https://github.com/HewlettPackard/oneview-chef/tree/master/examples/) (please include helpful comments).
- [ ] Changes documented in the [CHANGELOG](https://github.com/HewlettPackard/oneview-chef/blob/master/CHANGELOG.md).

### CODE REVIEW

File: spec/unit/resources/matchers_spec.rb [hunk 1]
```diff
@@ -41,6 +41,20 @@
...
+    expect(chef_run).to_not create_oneview_firmware_custom_spp('')
+
+    # oneview_connection_template
+    expect(chef_run).to_not update_oneview_connection_template('')
+    expect(chef_run).to_not reset_oneview_connection_template('')
+
+    # oneview_network_set
+    expect(chef_run).to_not create_oneview_network_set('')
+    expect(chef_run).to_not create_oneview_network_set_if_missing('')
+    expect(chef_run).to_not delete_oneview_network_set('')
```
Comment (jsmartt) [inline spec/unit/resources/matchers_spec.rb]: Not sure if you meant to add this in here or not, but it's actually defined below also (in alphabetical order)

File: README.md [hunk 2]
```diff
@@ -107,18 +107,28 @@ oneview_fcoe_network 'FCoE1' do
...
 ```
 
+```Ruby
+oneview_firmware 'CustomSPP'  do
+  client <my_client>
+  spp_name 'SPPName'
+  hotfixes_names [
+    'hotfix1_name'
+  ]
+  action :create_custom_spp
```
Comment (tmiotto) [inline README.md:133]: Where are the other actions?

File: resources/firmware.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    filename = ::File.basename(filename, ::File.extname(filename)).tr('.', '_')
+    firmware_list = OneviewSDK::FirmwareDriver.find_by(c, {})
+    firmware_list.find { |fw| filename == fw['resourceId'] }
+  end
+end
+
+action :add do
+  item = load_firmware
+
+  if item.nil?
```
Comment (tmiotto) [inline resources/firmware.rb]: Wouldn't make more sense replacing:

``` Ruby
if item.nil?
  do X
else
  do Y
end
```

for:

``` Ruby
if item
  do Y
else
  do X
end
```

???

File: resources/firmware.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    converge_by "Add #{resource_name} '#{name}'" do
+      OneviewSDK::FirmwareBundle.add(c, name)
+    end
+  else
+    Chef::Log.info "#{resource_name} '#{name}' is already up to date."
+  end
+end
+
+action :create_custom_spp do
+  raise "Unspecified property: 'spp_name' or 'spp_file'. Please set it before attempting this action." unless spp_name || spp_file
```
Comment (tmiotto) [inline resources/firmware.rb:53]: Both are required? Or just one of the variables?

File: resources/firmware.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  if hotfixes_names.empty? && hotfixes_files.empty?
+    raise "Unspecified property: 'hotfixes_names' or 'hotfixes_files'. Please set it before attempting this action."
```
Comment (tmiotto) [inline resources/firmware.rb:55]: You're checking if they are **empty** here, but the exception raised says they are **unspecified** (nil)

**Review by tmiotto — CHANGES_REQUESTED** [review]

File: resources/firmware.rb (hunk 3 above)
Comment (hdiomede) [reply resources/firmware.rb]: Yes, thanks!

File: resources/firmware.rb (hunk 4 above)
Comment (hdiomede) [reply resources/firmware.rb:53]: Only one

**Review by tmiotto — APPROVED** [review]

**Review by tmiotto — APPROVED** [review]

**Pull request merged**